[SPARK-28156][SQL][BACKPORT-2.4] Self-join should not miss cached view

bersprockets opened the pull request:
Back-port of #24960 to branch-2.4.

The issue is when self-join a cached view, only one side of join uses cached relation. The cause is in `ResolveReferences` we do deduplicate for a view to have new output attributes. Then in `AliasViewChild`, the rule adds extra project under a view. So it breaks cache matching.

The fix is when dedup, we only dedup a view which has output different to its child plan. Otherwise, we dedup on the view's child plan.

```scala
val df = Seq.tabulate(5) { x => (x, x + 1, x + 2, x + 3) }.toDF("a", "b", "c", "d")
df.write.mode("overwrite").format("orc").saveAsTable("table1")

sql("drop view if exists table1_vw")
sql("create view table1_vw as select * from table1")

val cachedView = sql("select a, b, c, d from table1_vw")

cachedView.createOrReplaceTempView("cachedview")
cachedView.persist()

val queryDf = sql(
  s"""select leftside.a, leftside.b
      |from cachedview leftside
      |join cachedview rightside
      |on leftside.a = rightside.a
    """.stripMargin)
```

Query plan before this PR:
```scala
== Physical Plan ==
*(2) Project [a#12664, b#12665]
+- *(2) BroadcastHashJoin [a#12664], [a#12660], Inner, BuildRight
   :- *(2) Filter isnotnull(a#12664)
   :  +- *(2) InMemoryTableScan [a#12664, b#12665], [isnotnull(a#12664)]
   :        +- InMemoryRelation [a#12664, b#12665, c#12666, d#12667], StorageLevel(disk, memory, deserialized, 1 replicas)
   :              +- *(1) FileScan orc default.table1[a#12660,b#12661,c#12662,d#12663] Batched: true, DataFilters: [], Format: ORC, Location: InMemoryF
ileIndex[file:/Users/viirya/repos/spark-1/sql/core/spark-warehouse/org.apache.spark.sql...., PartitionFilters: [], PushedFilters: [], ReadSchema: struc
t<a:int,b:int,c:int,d:int>
   +- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[0, int, true] as bigint)))
      +- *(1) Project [a#12660]
         +- *(1) Filter isnotnull(a#12660)
            +- *(1) FileScan orc default.table1[a#12660] Batched: true, DataFilters: [isnotnull(a#12660)], Format: ORC, Location: InMemoryFileIndex[fil
e:/Users/viirya/repos/spark-1/sql/core/spark-warehouse/org.apache.spark.sql...., PartitionFilters: [], PushedFilters: [IsNotNull(a)], ReadSchema: struc
t<a:int>
```

Query plan after this PR:
```scala
== Physical Plan ==
*(2) Project [a#12664, b#12665]
+- *(2) BroadcastHashJoin [a#12664], [a#12692], Inner, BuildRight
   :- *(2) Filter isnotnull(a#12664)
   :  +- *(2) InMemoryTableScan [a#12664, b#12665], [isnotnull(a#12664)]
   :        +- InMemoryRelation [a#12664, b#12665, c#12666, d#12667], StorageLevel(disk, memory, deserialized, 1 replicas)
   :              +- *(1) FileScan orc default.table1[a#12660,b#12661,c#12662,d#12663] Batched: true, DataFilters: [], Format: ORC, Location: InMemoryFileIndex[file:/Users/viirya/repos/spark-1/sql/core/spark-warehouse/org.apache.spark.sql...., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<a:int,b:int,c:int,d:int>
   +- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[0, int, false] as bigint)))
      +- *(1) Filter isnotnull(a#12692)
         +- *(1) InMemoryTableScan [a#12692], [isnotnull(a#12692)]
               +- InMemoryRelation [a#12692, b#12693, c#12694, d#12695], StorageLevel(disk, memory, deserialized, 1 replicas)
                     +- *(1) FileScan orc default.table1[a#12660,b#12661,c#12662,d#12663] Batched: true, DataFilters: [], Format: ORC, Location: InMemoryFileIndex[file:/Users/viirya/repos/spark-1/sql/core/spark-warehouse/org.apache.spark.sql...., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<a:int,b:int,c:int,d:int>
```

Added test.

viirya committed:
[SPARK-28156][SQL] Self-join should not miss cached view

The issue is when self-join a cached view, only one side of join uses cached relation. The cause is in `ResolveReferences` we do deduplicate for a view to have new output attributes. Then in `AliasViewChild`, the rule adds extra project under a view. So it breaks cache matching.

The fix is when dedup, we only dedup a view which has output different to its child plan. Otherwise, we dedup on the view's child plan.

```scala
val df = Seq.tabulate(5) { x => (x, x + 1, x + 2, x + 3) }.toDF("a", "b", "c", "d")
df.write.mode("overwrite").format("orc").saveAsTable("table1")

sql("drop view if exists table1_vw")
sql("create view table1_vw as select * from table1")

val cachedView = sql("select a, b, c, d from table1_vw")

cachedView.createOrReplaceTempView("cachedview")
cachedView.persist()

val queryDf = sql(
  s"""select leftside.a, leftside.b
      |from cachedview leftside
      |join cachedview rightside
      |on leftside.a = rightside.a
    """.stripMargin)
```

Query plan before this PR:
```scala
== Physical Plan ==
*(2) Project [a#12664, b#12665]
+- *(2) BroadcastHashJoin [a#12664], [a#12660], Inner, BuildRight
   :- *(2) Filter isnotnull(a#12664)
   :  +- *(2) InMemoryTableScan [a#12664, b#12665], [isnotnull(a#12664)]
   :        +- InMemoryRelation [a#12664, b#12665, c#12666, d#12667], StorageLevel(disk, memory, deserialized, 1 replicas)
   :              +- *(1) FileScan orc default.table1[a#12660,b#12661,c#12662,d#12663] Batched: true, DataFilters: [], Format: ORC, Location: InMemoryF
ileIndex[file:/Users/viirya/repos/spark-1/sql/core/spark-warehouse/org.apache.spark.sql...., PartitionFilters: [], PushedFilters: [], ReadSchema: struc
t<a:int,b:int,c:int,d:int>
   +- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[0, int, true] as bigint)))
      +- *(1) Project [a#12660]
         +- *(1) Filter isnotnull(a#12660)
            +- *(1) FileScan orc default.table1[a#12660] Batched: true, DataFilters: [isnotnull(a#12660)], Format: ORC, Location: InMemoryFileIndex[fil
e:/Users/viirya/repos/spark-1/sql/core/spark-warehouse/org.apache.spark.sql...., PartitionFilters: [], PushedFilters: [IsNotNull(a)], ReadSchema: struc
t<a:int>
```

Query plan after this PR:
```scala
== Physical Plan ==
*(2) Project [a#12664, b#12665]
+- *(2) BroadcastHashJoin [a#12664], [a#12692], Inner, BuildRight
   :- *(2) Filter isnotnull(a#12664)
   :  +- *(2) InMemoryTableScan [a#12664, b#12665], [isnotnull(a#12664)]
   :        +- InMemoryRelation [a#12664, b#12665, c#12666, d#12667], StorageLevel(disk, memory, deserialized, 1 replicas)
   :              +- *(1) FileScan orc default.table1[a#12660,b#12661,c#12662,d#12663] Batched: true, DataFilters: [], Format: ORC, Location: InMemoryFileIndex[file:/Users/viirya/repos/spark-1/sql/core/spark-warehouse/org.apache.spark.sql...., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<a:int,b:int,c:int,d:int>
   +- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[0, int, false] as bigint)))
      +- *(1) Filter isnotnull(a#12692)
         +- *(1) InMemoryTableScan [a#12692], [isnotnull(a#12692)]
               +- InMemoryRelation [a#12692, b#12693, c#12694, d#12695], StorageLevel(disk, memory, deserialized, 1 replicas)
                     +- *(1) FileScan orc default.table1[a#12660,b#12661,c#12662,d#12663] Batched: true, DataFilters: [], Format: ORC, Location: InMemoryFileIndex[file:/Users/viirya/repos/spark-1/sql/core/spark-warehouse/org.apache.spark.sql...., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<a:int,b:int,c:int,d:int>
```

Added test.

Closes #24960 from viirya/SPARK-28156.

Authored-by: Liang-Chi Hsieh <viirya@gmail.com>
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

SparkQA commented:
**[Test build #107322 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107322/testReport)** for PR 25068 at commit [`f2645aa`](https://github.com/apache/spark/commit/f2645aa28e99577dfed9645337cce17376cf53a5).

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

dongjoon-hyun commented:
cc @cloud-fan and @viirya

viirya commented:
Thanks for pinging me. Will take a look tonight.

On Mon, Jul 8, 2019, 14:37 Dongjoon Hyun <notifications@github.com> wrote:

> cc @cloud-fan <https://github.com/cloud-fan> and @viirya
> <https://github.com/viirya>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/spark/pull/25068?email_source=notifications&email_token=AAAQZ57JESRH4ZMFIFA3LITP6LOBBA5CNFSM4H6XNRIKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZMDO3I#issuecomment-509097837>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAQZ553CXLNQUFNV5U6FCLP6LOBBANCNFSM4H6XNRIA>
> .
>

SparkQA commented:
**[Test build #107322 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107322/testReport)** for PR 25068 at commit [`f2645aa`](https://github.com/apache/spark/commit/f2645aa28e99577dfed9645337cce17376cf53a5).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107322/
Test FAILed.

dongjoon-hyun commented:
Retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12474/
Test PASSed.

SparkQA commented:
**[Test build #107332 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107332/testReport)** for PR 25068 at commit [`f2645aa`](https://github.com/apache/spark/commit/f2645aa28e99577dfed9645337cce17376cf53a5).

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:
```diff
@@ -148,6 +148,40 @@ object Cast {
+   * precision lose or possible runtime failures. For example, long -> int, string -> int are not
+   * up-cast.
+   */
+  def canUpCast(from: DataType, to: DataType): Boolean = (from, to) match {
```
Where is this from? @bersprockets Can you note the original PR of this in the PR description?

viirya replied:
I think this should have some tests in the original PR. Should we backport them together? @cloud-fan @dongjoon-hyun

dongjoon-hyun replied:
Usually, we don't backport the original PR if this kind of requirement arises. How do you think about this backport, @cloud-fan and @gatorsmile ?

This PR needs `canUpCast` (here) and `sameOutput` (https://github.com/apache/spark/pull/25068/files#diff-27c76f96a7b2733ecfd6f46a1716e153R140) additionally.

cloud-fan replied:
The rule `AliasViewChild` is already there before we have `canUpCast` and `sameOutput`. We just need to update the corresponding code to the old version without `canUpCast` and `sameOutput`.

dongjoon-hyun replied:
+1 for @cloud-fan 's comment. Thanks!

viirya commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala:
```diff
@@ -604,4 +606,25 @@ class AnalysisSuite extends AnalysisTest with Matchers {
     }
   }
+
+  test("SPARK-25691: AliasViewChild with different nullabilities") {
```
hmm, this test actually comes from another PR`SPARK-25691`. So the original PR doesn't backport too?

bersprockets replied:
It got pulled in by the cherry pick because your original PR patched it, and I didn't see harm in the added test. However, its title references a patch that was not pulled in, which is weird, so I should probably remove it.

dongjoon-hyun replied:
Yep. Please remove this `test("SPARK-25691...")` since this is SPARK-28156 backport.

viirya reviewed (commented):
This backport looks good. It contains more than original PR, I'm not sure if it is ok to include the changes of other PRs in a backport PR.

SparkQA commented:
**[Test build #107332 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107332/testReport)** for PR 25068 at commit [`f2645aa`](https://github.com/apache/spark/commit/f2645aa28e99577dfed9645337cce17376cf53a5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107332/
Test PASSed.

bersprockets commented:
@viirya 

Thanks for the comments. I was a little perplexed about the case where a back-ported PR needed a few small utility methods from other PRs: should I completely back-port the other PRs, or should I just include the two needed utility methods?

I chose to copy in the two utility methods. However, I will follow whatever is the standard.

(The two other PRs are not included in the title as they are not actually back-ported).

Cast::canUpCast comes from SPARK-24586 ("Upcast should not allow casting from string to other types"). canUpCast is referenced by SPARK-28156 in CheckAnalysis. 

LogicalPlan::sameOutput comes from SPARK-25691 ("Use semantic equality in AliasViewChild in order to compare attributes"). sameOutput is referenced by SPARK-28156 in CheckAnalysis.

bersprockets committed:
Remove extraneous test

SparkQA commented:
**[Test build #107365 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107365/testReport)** for PR 25068 at commit [`ca292d8`](https://github.com/apache/spark/commit/ca292d88e826467cf2acf63fb038f0bd0d6a6ded).

SparkQA commented:
**[Test build #107365 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107365/testReport)** for PR 25068 at commit [`ca292d8`](https://github.com/apache/spark/commit/ca292d88e826467cf2acf63fb038f0bd0d6a6ded).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107365/
Test FAILed.

dongjoon-hyun commented:
Retest this please.

dongjoon-hyun commented:
The R failure is irrelevant to this PR.
```
* checking CRAN incoming feasibility ...Error in readRDS(con) : error reading from connection
Execution halted
```

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12515/
Test PASSed.

SparkQA commented:
**[Test build #107377 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107377/testReport)** for PR 25068 at commit [`ca292d8`](https://github.com/apache/spark/commit/ca292d88e826467cf2acf63fb038f0bd0d6a6ded).

SparkQA commented:
**[Test build #107377 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107377/testReport)** for PR 25068 at commit [`ca292d8`](https://github.com/apache/spark/commit/ca292d88e826467cf2acf63fb038f0bd0d6a6ded).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107377/
Test PASSed.

dongjoon-hyun commented:
Hi, @bersprockets .
For backporting, could you follow @cloud-fan 's [advice](https://github.com/apache/spark/pull/25068#discussion_r301416338)?
> The rule AliasViewChild is already there before we have canUpCast and sameOutput. We just need to update the corresponding code to the old version without canUpCast and sameOutput.

bersprockets commented:
@dongjoon-hyun @cloud-fan will do. Thanks.

dongjoon-hyun commented:
Gentle ping, @bersprockets .

bersprockets commented:
@dongjoon-hyun Thanks :). I am away but will return to it next week.

bersprockets committed:
Remove extraneous methods

bersprockets commented:
I did the following:

- replaced `!v.sameOutput(child)` with `output != child.output`
- replaced `!Cast.canUpCast` with `Cast.mayTruncate`

In the process, I broke a test in SQLViewSuite. I will hunt down the cause tomorrow and hopefully post the changes.

dongjoon-hyun commented:
Thank you for informing that, @bersprockets !

SparkQA commented:
**[Test build #108138 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108138/testReport)** for PR 25068 at commit [`474c16e`](https://github.com/apache/spark/commit/474c16ea5eae33da3f198c8f240a77ae8b1ce13e).

SparkQA commented:
**[Test build #108138 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108138/testReport)** for PR 25068 at commit [`474c16e`](https://github.com/apache/spark/commit/474c16ea5eae33da3f198c8f240a77ae8b1ce13e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108138/
Test PASSed.

dongjoon-hyun commented:
Could you review this please, @cloud-fan ?

cloud-fan commented:
thanks, merging to 2.4!

dongjoon-hyun commented:
Thank you, @cloud-fan and @bersprockets !

bersprockets commented:
Thanks for the help everyone.

maropu commented:
ping @bersprockets

bersprockets commented:
@maropu I got pinged on this PR but now can't find the pinging comment. I hope I didn't break anything.

maropu commented:
NVM, that's my mistake.
